libcudacxx/include/cuda/std/__type_traits/type_set.h

```diff
@@ -0,0 +1,92 @@
+//===----------------------------------------------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+// SPDX-FileCopyrightText: Copyright (c) 2023 NVIDIA CORPORATION & AFFILIATES.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef _LIBCUDACXX___TYPE_TRAITS_TYPE_SET_H
+#define _LIBCUDACXX___TYPE_TRAITS_TYPE_SET_H
+
+#include <cuda/std/detail/__config>
+
+#if defined(_CCCL_IMPLICIT_SYSTEM_HEADER_GCC)
+#  pragma GCC system_header
+#elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_CLANG)
+#  pragma clang system_header
+#elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_MSVC)
+#  pragma system_header
+#endif // no system header
+
+#include <cuda/std/__concepts/__concept_macros.h>
+#include <cuda/std/__concepts/all_of.h>
+#include <cuda/std/__type_traits/conditional.h>
+#include <cuda/std/__type_traits/is_base_of.h>
+#include <cuda/std/__type_traits/type_identity.h>
+#include <cuda/std/cstddef>
+
+_LIBCUDACXX_BEGIN_NAMESPACE_STD
+
+#if __has_builtin(__is_base_of) || defined(_LIBCUDACXX_MSVC)
+template <class _Set, class... _Ty>
+_LIBCUDACXX_CONCEPT __type_set_contains = _CUDA_VSTD::__all_of<__is_base_of(type_identity<_Ty>, _Set)...>;
+#else
+template <class _Set, class... _Ty>
+_LIBCUDACXX_CONCEPT __type_set_contains =
+  _CUDA_VSTD::__all_of<_CCCL_TRAIT(_CUDA_VSTD::is_base_of, type_identity<_Ty>, _Set)...>;
+#endif
```

Comment:
I do not follow, the trait assigns to a single inline variable, and uses that. How is that measurably slower than just the buildin

Reply:
The variable template traits are specified to access the static `value` member of the associated class template. So they cause the instantiation of a class template, unless they are non-conforming or there is compiler heroics to avoid the cost (msvc).

Reply:
BTW this is also the same for all our implementations. Whenever we have the chance to use a builtin, then we directly access it. So we do not rely on another STL here